discovery-provider/src/queries/get_notifications.py

```diff
@@ -0,0 +1,221 @@
+  n.group_id, user_seen.seen_at, user_seen.prev_seen_at
+ORDER BY
+  user_seen.seen_at desc NULLS LAST,
+  n.group_id desc
+limit :limit;
+"""
+)
+
+
+MAX_LIMIT = 50
+DEFAULT_LIMIT = 20
+
+
+class NotificationGroup(TypedDict):
+    group_id: str
+    notification_ids: List[int]
+    is_seen: bool
+    seen_at: datetime
+    prev_seen_at: datetime
+    count: int
+
+
+def get_notification_groups(session: Session, args: GetNotificationArgs):
+    """
+    Gets the user's notifiations in the database
+    """
+
+    limit = min(args.get("limit", DEFAULT_LIMIT), MAX_LIMIT)  # type: ignore
+
+    rows = session.execute(
+        notification_groups_sql,
+        {
+            "user_id": args["user_id"],
+            "limit": limit,
+            "timestamp_offset": None,
+            "group_id_offset": None,
+        },
+    )
+    res: List[NotificationGroup] = [
+        {
+            "group_id": r[0],
+            "notification_ids": r[1],
+            "is_seen": r[2] if r[2] != None else False,
+            "seen_at": r[3],
+            "prev_seen_at": r[4],
+            "count": r[5],
+        }
+        for r in rows
+    ]
+    logger.info(res)
+
+    return res
+
+
+class FollowNotification(TypedDict):
+    follower_user_id: int
+    followee_user_id: int
+
+
+class RepostNotification(TypedDict):
+    type: str
+    user_id: int
+    repost_item_id: int
+
+
+class SaveNotification(TypedDict):
+    type: str
+    user_id: int
+    save_item_id: int
+
+
+class MilestoneNotification(TypedDict):
+    type: str
+    threshold: int
+
+
+class RemixNotification(TypedDict):
+    parent_track_id: int
+    track_id: int
+
+
+class CosignRemixNotification(TypedDict):
+    parent_track_id: int
+    track_id: int
+
+
+NotificationData = Union[
+    FollowNotification,
+    RepostNotification,
+    SaveNotification,
+    MilestoneNotification,
+    RemixNotification,
+    CosignRemixNotification,
+]
+
+
+class NotificationAction(TypedDict):
+    specifier: str
+    type: str
+    timestamp: datetime
+    data: NotificationData
+
+
+class Notification(TypedDict):
+    group_id: str
+    is_seen: bool
+    seen_at: datetime
+    actions: List[NotificationAction]
+
+
+notifications_sql = text(
+    """
+SELECT
+    id,
+    type,
+    specifier,
+    timestamp,
+    data
+FROM notification n
+WHERE n.id in :notification_ids
+"""
+)
+notifications_sql = notifications_sql.bindparams(
+    bindparam("notification_ids", expanding=True)
+)
+
+
+def get_notifications(session: Session, args: GetNotificationArgs):
+    notifications = get_notification_groups(session, args)
+    notification_ids = []
+    logger.info(notifications)
+    for notification in notifications:
+        notification_ids.extend(notification["notification_ids"])
+
+    rows = session.execute(notifications_sql, {"notification_ids": notification_ids})
+
+    notification_id_data: Dict[int, NotificationAction] = defaultdict()
+
+    for row in rows:
+        notification_id_data[row[0]] = {
```

Comment:
Is it possible to destructure an arrest into variables? Might be helpful to name the values from row first?